### ISSUE TITLE
[Question] Why use `__call__` to do forward.

### ISSUE THREAD

**Issue by Felix-Zhenghao** [karpathy/nanoGPT#475]
In the <a href="https://github.com/karpathy/nanoGPT/blob/master/model.py#L316">code</a> of `model.py`, the inference part, the author doesn't use the dot syntax to access the forward method of class `GPT`:
```Python
logits, _ = self(idx_cond)
```

Why not use `logits, _ = self.forward(idx_cond)`? Is there any further consideration related to the `__call__` method of `nn.Module`? If not, I believe the code should be changed to `logits, _ = self.forward(idx_cond)` for readability.

**Comment by ruipreis** [comment]
This is related to PyTorch's processing flow. The `__call__` allows for hooks. You can read more about it here:

https://stackoverflow.com/questions/55338756/why-there-are-different-output-between-model-forwardinput-and-modelinput

**Comment by Felix-Zhenghao** [comment]
Thanks @ruipreis
